fix: Add Keycloak health port 9000 starting from major version 25

paulomorgado opened the pull request:
## What does this PR do?

Starting with tag `25.0` Keycloak uses port 9000 instead of port 8080 for health checks.

## Why is it important?

Currently the readiness for image `25.0` and above is not correct and might consider the container ready when it's yet ready.

## Related issues

- Closes #1208

## How to test this PR

Use tag 24.0 and tag 25.0 and asset that all correctly wait for the container to be ready.

paulomorgado committed:
Enhance Keycloak container setup

- Added `KeycloakHealthPort` constant set to 9000 in `KeycloakBuilder` for health check configurations.
- Introduced `DefaultUsername` and `DefaultPassword` constants as "admin" in `KeycloakBuilder` for default access credentials.
- Implemented logic for special handling of Keycloak version 25 and above, including:
  * Parsing the major version from the image tag.
  * Configuring `KeycloakHealthPort` port binding and a wait strategy based on the `/health/ready` endpoint for versions 25+, adapting to new health check mechanisms.

paulomorgado committed:
Refactor KeycloakBuilder and add health tests

This commit makes significant updates to the KeycloakBuilder and introduces a new test class for Keycloak container health. In the KeycloakBuilder, the Build method has been revised to remove the immediate call to `Validate()` and the direct return of a new `KeycloakContainer`. Instead, it now includes logic to extract the major version number from the `DockerResourceConfiguration.Image.Tag`. If the version is 25 or higher, the builder is modified to add port binding for the Keycloak health port and a wait strategy for the `/health/ready` endpoint. After these potential modifications, `Validate()` is called to ensure the configuration is correct before returning the container.

Additionally, the validation logic that previously checked the Docker image tag's major version has been removed from the direct validation method and incorporated into the Build method. This change streamlines the validation process, focusing it solely on configuration validation.

A new test class, `KeycloakContainerHealthTest`, has been added to verify the health of the Keycloak container. It includes a method `HealthIsReady` that tests whether the container starts successfully and if its health endpoint returns an HTTP OK status. This method dynamically adjusts the `HttpClient.BaseAddress` based on the container's accessible hostname and port, ensuring the test is robust across different environments. The test covers both below and at the threshold versions with inline data for "quay.io/keycloak/keycloak:24.0" and "quay.io/keycloak/keycloak:25.0", ensuring the new health check configuration is properly applied. Containers are disposed of post-test to prevent resource leakage.

HofmeisterAn commented:
For reference, it looks like the changes were introduced in version 25, according to the [release notes](https://www.keycloak.org/docs/latest/release_notes/index.html#management-port-for-metrics-and-health-endpoints).

HofmeisterAn committed:
Merge branch 'develop' into keycloack-health

HofmeisterAn commented on src/Testcontainers.Keycloak/KeycloakBuilder.cs:
```diff
@@ -59,8 +61,21 @@ public KeycloakBuilder WithPassword(string password)
+        }
+
+        builder.Validate();
+        return new KeycloakContainer(builder.DockerResourceConfiguration);
```
```suggestion
        Validate();

        var majorVersionString = new string(DockerResourceConfiguration.Image.Tag.TakeWhile(char.IsDigit).ToArray());

        var isLatestVersion = "latest".Equals(DockerResourceConfiguration.Image.Tag, StringComparison.OrdinalIgnoreCase);

        // https://www.keycloak.org/docs/latest/release_notes/index.html#management-port-for-metrics-and-health-endpoints.
        var isMajorVersionAtLeast25 = int.TryParse(majorVersionString, out var majorVersion) && majorVersion >= 25;

        var waitStrategy = Wait.ForUnixContainer()
            .UntilHttpRequestIsSucceeded(request =>
                request.ForPath("/health/ready").ForPort(isLatestVersion || isMajorVersionAtLeast25 ? KeycloakHealthPort : KeycloakPort));

        var couchbaseBuilder = DockerResourceConfiguration.WaitStrategies.Count() > 1 ? this : WithWaitStrategy(waitStrategy);
        return new KeycloakContainer(couchbaseBuilder.DockerResourceConfiguration);
```

We need to check the `latest` tag as well. In addition, the tag does not necessarily require a dot separator. I am not sure if this is the case for the Keycloak images too, but sometimes tags for images are like: 25, 25.1, 25.1.3, etc. This requires updating the global usings as well (add LINQ).

In addition we can remove the wait strategy configuration from `Init()`.

HofmeisterAn commented on tests/Testcontainers.Keycloak.Tests/KeycloakContainerHealthTest.cs:
```diff
@@ -0,0 +1,41 @@
+
+    [Theory]
+    [InlineData("quay.io/keycloak/keycloak:24.0")]
+    [InlineData("quay.io/keycloak/keycloak:25.0")]
```
In case we need to run the same tests against different configurations (versions), we usually use the following pattern (this allows us to run the tests and download the images in parallel). For example, this is from [Mongo DB](https://github.com/testcontainers/testcontainers-dotnet/blob/21137c61fb36774b0aee16304e1587fa0f662673/tests/Testcontainers.MongoDb.Tests/MongoDbContainerTest.cs#L52-L86). Would you mind if we change the test to follow the same pattern?

paulomorgado replied:
I don't understand what's being asked here.

But, if you guide me, I'm sure I can do it.

I don't use JetBrains products.

HofmeisterAn reviewed (changes requested):
Sorry it took me so long to review. Thank you for taking the time to contribute, create a PR, and fix the issue. I have a few small change requests. What do you think? Thanks again.

paulomorgado committed:
Update src/Testcontainers.Keycloak/KeycloakBuilder.cs

PR suggestion

Co-authored-by: Andre Hofmeister <9199345+HofmeisterAn@users.noreply.github.com>

paulomorgado committed:
Improve KeycloakBuilder version handling and consistency

Enhanced `KeycloakBuilder` to better handle the `isLatestVersion` flag by including "nightly" tags and refining major version parsing. Updated the wait strategy to use `KeycloakHealthPort` for the latest versions. Renamed `couchbaseBuilder` to `keycloakBuilder` for consistency. Adjusted wait strategy condition to check for any strategies. Added `System.Linq` to global usings in `Usings.cs`.

paulomorgado committed:
Improve KeycloakBuilder version handling and consistency

Enhanced `KeycloakBuilder` to better handle the `isLatestVersion` flag by including "nightly" tags and refining major version parsing. Updated the wait strategy to use `KeycloakHealthPort` for the latest versions. Renamed `couchbaseBuilder` to `keycloakBuilder` for consistency. Adjusted wait strategy condition to check for any strategies. Added `System.Linq` to global usings in `Usings.cs`.

HofmeisterAn committed:
feat: Add IImage.MatchVersion

HofmeisterAn committed:
chore: Use new IImage.MatchVersion member

HofmeisterAn committed:
Merge branch 'keycloack-health' of github.com:paulomorgado/testcontainers-testcontainers-dotnet into paulomorgado-keycloack-health

HofmeisterAn committed:
docs: Add XML doc

HofmeisterAn commented:
Thanks @paulomorgado 🙏.
